python/lsst/daf/butler/registry/collections/_base.py

```diff
@@ -0,0 +1,485 @@
+        for child, restriction in children.iter(manager, withRestrictions=True, flattenChains=False):
+            if restriction.names is ...:
+                rows.append({"parent": self._id, "child": child.key,
+                             "position": next(position), "dataset_type_name": ""})
+            else:
+                for name in restriction.names:
+                    rows.append({"parent": self._id, "child": child.key,
+                                 "position": next(position), "dataset_type_name": name})
+        with self._db.transaction():
+            self._db.delete(self._table, ["parent"], {"parent": self._id})
+            self._db.insert(self._table, *rows)
+
+    def _load(self, manager: CollectionManager) -> CollectionSearch:
+        # Docstring inherited from ChainedCollectionRecord.
+        sql = sqlalchemy.sql.select(
+            [self._table.columns.child, self._table.columns.dataset_type_name]
+        ).select_from(
+            self._table
+        ).where(
+            self._table.columns.parent == self.key
+        ).order_by(
+            self._table.columns.position
+        )
+        # It's fine to have consecutive rows with the same collection name
+        # and different dataset type names - CollectionSearch will group those
+        # up for us.
+        children = []
+        for row in self._db.query(sql):
+            key = row[self._table.columns.child]
+            restriction = row[self._table.columns.dataset_type_name]
+            if not restriction:
+                restriction = ...  # we store ... as "" in the database
+            record = manager[key]
+            children.append((record.name, restriction))
+        return CollectionSearch.fromExpression(children)
+
+
+class DefaultCollectionManager(CollectionManager):
+    """Default `CollectionManager` implementation.
+
+    This implementation uses record classes defined in this module and is
+    based on the same assumptions about schema outlined in the record classes.
+
+    Parameters
+    ----------
+    db : `Database`
+        Interface to the underlying database engine and namespace.
+    tables : `NamedTuple`
+        Named tuple of SQLAlchemy table objects.
+    collectionIdName : `str`
+        Name of the column in collections table that identifies it (PK).
+
+    Note
+    ----
+    Implementation uses "aggressive" pre-fetching and caching of the records
+    in memory. Memory cache is synchronized from database when `refresh`
+    method is called.
+    """
+    def __init__(self, db: Database, tables: NamedTuple[sqlalchemy.schema.Table, ...],
+                 collectionIdName: str):
+        self._db = db
+        self._tables = tables
+        self._collectionIdName = collectionIdName
+        self._records = {}  # indexed by record ID
+
+    @classmethod
+    def getCollectionForeignKeyName(cls, prefix: str = "collection") -> str:
+        # Docstring inherited from CollectionManager.
+        return f"{prefix}_id"
```

Comment:
I'd prefer to leave this method and the one below it abstract, I think, and let subclasses implement the suffix to be `_id` or `_name` to match the primary key.  We have a bit of an informal convention (or at least a lot precedent) for `id` fields being integers only.